core: support to create pool with assigned pg and can set 'autoscale mode' (#11967)
rook/rook: Documentation/CRDs/Block-Storage/ceph-block-pool-crd.md

    !!! note
        A value of 0 disables the quota.

> * `placementGroupCount`: Set a custom PG count for the pool. By default, the PGs will be managed by the Ceph PG autoscaler. See the [Ceph documentation](https://docs.ceph.com/en/latest/rados/operations/pools/#create-a-pool) for more info.

Reviewer: Do we need this setting? The user can currently set the pg count with this, right? Any pool setting found [here](https://docs.ceph.com/en/latest/rados/operations/pools/#set-pool-values) can be set with the `parameters`.
```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: replicapool
  namespace: rook-ceph # namespace:cluster
spec:
  parameters:
    pg_num: "32"
    nopgchange: "1"
```
I'm not sure if the `nopgchange` is the same as disabling the autoscaler, so perhaps the `autoscaleMode` would still be required in the schema. 

Where we can find a way to use the `parameters`, it's nice that we wouldn't need to update the schema.

Author: In summary, nopgchange and autoscalemode are both settings that can be used in Ceph to control the scaling behavior of the storage cluster, but they operate at different levels and have different effects. nopgchange is used to control whether individual PGs can be moved between OSDs, while autoscalemode is used to control how the cluster automatically scales its resources based on the workload.

Author: maybe we can add "pg_nm" and "autoscalemode" to the **parameters**, no need to add to the schema?

Reviewer 2: I see documentation for `pg_autoscale_mode` here: https://docs.ceph.com/en/latest/rados/operations/placement-groups/#autoscaling-placement-groups

Reviewer 2: @zhucan please stop ongoing conversations as resolved on GitHub. It takes ages to find where past conversations have been started. Instead, please respond to the question.

> * `pgs`: Set pg numbers to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/pools/#create-a-pool) for more info. If the autosacale mode is on, the pg numbers will be set to default value, the assigned value maybe invalid.

Reviewer: Since autoscale is always on, here is a suggestion on the wording. 
```suggestion
* `pgs`: Set a custom PG count for the pool. By default, the PGs will be managed by the Ceph PG autoscaler. See the [Ceph documentation](https://docs.ceph.com/en/latest/rados/operations/pools/#create-a-pool) for more info. 
```

> * `autoscaleMode`: set autoscale mode to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.

Author: Maybe warn is suitable，if the pg is not enough，we should print it to custom

> * `autoscaleMode`: set autoscale mode to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.

Author: If the default value if off，there is no any other warning message for it，so the default is wan is suitable，I think

> * `autoscaleMode`: set autoscale mode to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.

Reviewer 2: @zhucan please respond.

> * `autoscaleMode`: set autoscale mode to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.

Author: @travisn Because of may we will scale the ceph cluster and the pg_num is not enough, we need the warnning message if the pg is not enough.

    * `off`: Disable autoscaling for this pool. It is up to the administrator to choose an appropriate pgp_num for each pool. Please refer to [Choosing the number of Placement Groups](https://docs.ceph.com/en/latest/rados/operations/placement-groups/#choosing-number-of-placement-groups) for more information.
    * `on`: Enable automated adjustments of the PG count for the given pool.
    * `warn`: Raise health alerts when the PG count should be adjusted.

> 

Reviewer 2: Should we document that setting this to `""` (or leaving it unset) reverts to Ceph's default behavior?

Author: no, if  the PlacementGroupCount > 0, we should set the autoscalemode to warn or off in default. if PlacementGroupCount not set, the default value of the autoscalemode  should be on.

Reviewer 2: But Ceph should have its own default behavior. Why can we not just let ceph handle its own defaults.

Author: If user set the pg_num means want to disable the autoscalemode, otherwise it doesn't make any sense.

Reviewer 2: You are correct, but seemed to be missing my intent. Ceph often has its own behavior relating to configurations, and it is often best for Rook to rely on Ceph behaviors when they exist. In this case, after inspecting Ceph's C++ code, I did not find any Ceph behaviors that already accomplish what is being implemented here, so it is appropriate for Rook to do so.

    !!! note
        A value of 0 disables the quota.

> * `autoscaleMode`: set autoscale mode to pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.

Reviewer: ```suggestion
* `autoscaleMode`: Set the autoscale mode for the pool. See the [ceph documentation](https://docs.ceph.com/en/latest/rados/operations/placement-groups/) for more info.
```

    * `off`: Disable autoscaling for this pool. It is up to the administrator to choose an appropriate pgp_num for each pool. Please refer to [Choosing the number of Placement Groups](https://docs.ceph.com/en/latest/rados/operations/placement-groups/#choosing-number-of-placement-groups) for more information.
    * `on`: Enable automated adjustments of the PG count for the given pool.
    * `warn`: Raise health alerts when the PG count should be adjusted.

> 

Reviewer 2: I find that it is best to also document the value that will implement defaulting behavior.

```suggestion
    * `` (empty): Use the default Ceph-configured pool autoscale strategy.

```
